src/stream/src/executor/backfill/utils.rs

```diff
@@ -79,32 +89,147 @@ impl BackfillState {
     pub(crate) fn update_progress(
         &mut self,
         vnode: VirtualNode,
-        progress: BackfillProgressPerVnode,
-    ) -> Option<BackfillProgressPerVnode> {
-        self.inner.insert(vnode, progress)
+        new_pos: OwnedRow,
+    ) -> StreamExecutorResult<()> {
+        let state = self.get_current_state(&vnode);
+        let new_state = BackfillProgressPerVnode::InProgress(new_pos);
+        match state {
+            BackfillProgressPerVnode::NotStarted => *state = new_state,
+            BackfillProgressPerVnode::InProgress(_current_pos) => *state = new_state,
+            BackfillProgressPerVnode::Completed { .. } => unreachable!(),
+        }
+        Ok(())
     }
 
-    fn iter_backfill_progress(
-        &self,
-    ) -> impl Iterator<Item = (&VirtualNode, &BackfillProgressPerVnode)> {
-        self.inner.iter()
+    pub(crate) fn finish_progress(&mut self, vnode: VirtualNode, pos_len: usize) {
+        let finished_placeholder_position = construct_initial_finished_state(pos_len);
+        let current_state = self.get_current_state(&vnode);
+        let new_pos = match current_state {
+            BackfillProgressPerVnode::NotStarted => finished_placeholder_position,
+            BackfillProgressPerVnode::InProgress(current_pos) => current_pos.clone(),
+            BackfillProgressPerVnode::Completed { .. } => {
+                return;
+            }
+        };
+        *current_state = BackfillProgressPerVnode::Completed(new_pos);
+    }
+
+    /// If can commit, return state to be committed.
+    fn get_commit_state(&self, vnode: &VirtualNode) -> Option<(Option<Vec<Datum>>, Vec<Datum>)> {
+        if !self.can_commit(vnode) {
+            return None;
+        }
+        let new_state = self.inner.get(vnode).unwrap().current_state().clone();
+        let new_encoded_state = match new_state {
+            BackfillProgressPerVnode::NotStarted => unreachable!(),
+            BackfillProgressPerVnode::InProgress(current_pos) => {
+                let mut encoded_state = vec![None; current_pos.len() + METADATA_STATE_LEN];
+                encoded_state[0] = Some(vnode.to_scalar().into());
+                encoded_state[1..current_pos.len() + 1].clone_from_slice(current_pos.as_inner());
+                encoded_state[current_pos.len() + 1] = Some(false.into());
+                encoded_state[current_pos.len() + 2] = Some(0i64.into());
```

Comment:
The `row_count` field is always zero? It seems we should pass `total_snapshot_processed_rows` here.

Reply:
I will refactor it separately. https://github.com/risingwavelabs/risingwave/issues/10225.

Requires changes to the vnode level state.